csharp/ql/src/semmle/code/csharp/dataflow/internal/rangeanalysis/RangeUtils.qll

```diff
@@ -0,0 +1,17 @@
+private import csharp
+private import Ssa
+private import SsaReadPositionCommon
+private import semmle.code.csharp.controlflow.Guards as G
+
+private class Guard = G::Guard;
+
+private class BooleanValue = G::AbstractValues::BooleanValue;
+
+/**
+ * Holds if `guard` controls the position `controlled` with the value `testIsTrue`.
+ */
+predicate guardControlsSsaRead(Guard guard, SsaReadPosition controlled, boolean testIsTrue) {
+  exists(BooleanValue b | b.getValue() = testIsTrue |
```

Comment:
This is changed from Java, a probably is a lot more simplified than it should be.

Reply:
Let my try to explain some of the subtleties involved with the implies logic as it exists in the Java libs.  In general we want to apply as much implies logic as possible to the general guard-controls relation.  There are basic implications that merely rely on (base) SSA, such as assignment of the guard to a boolean variable that is subsequently checked:
```
boolean b = guard;
if (b) { /* guarded */ }
```
There are also more complicated implications that rely on range analysis.  So in the range analysis libraries we want to use as many implications as possible, but we can't use the full set, as that would tie a very big recursive knot (possibly with negative recursion).